File: App_frontend/app/src/main/java/com/mike/mikeapp/model/Card.java
package com.mike.mikeapp.model;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;
import com.google.gson.annotations.SerializedName;
import java.io.Serializable;

/**
 * 名片数据模型类
 * 对应后端API的25个标准字段
 */
@Entity(tableName = "cards")
public class Card implements Serializable {
    @PrimaryKey(autoGenerate = true)
    private int id;

    // 基本信息 (8字段)
    @ColumnInfo(name = "name_zh")
    @SerializedName("name_zh")
    private String nameZh;

    @ColumnInfo(name = "name_en")
    @SerializedName("name_en")
    private String nameEn;

    @ColumnInfo(name = "company_name_zh")
    @SerializedName("company_name_zh")
    private String companyNameZh;

    @ColumnInfo(name = "company_name_en")
    @SerializedName("company_name_en")
    private String companyNameEn;

    @ColumnInfo(name = "position_zh")
    @SerializedName("position_zh")
    private String positionZh;

    @ColumnInfo(name = "position_en")
    @SerializedName("position_en")
    private String positionEn;

    @ColumnInfo(name = "position1_zh")
    @SerializedName("position1_zh")
    private String position1Zh;

    @ColumnInfo(name = "position1_en")
    @SerializedName("position1_en")
    private String position1En;

    // 组织架构 (6字段)
    @ColumnInfo(name = "department1_zh")
    @SerializedName("department1_zh")
    private String department1Zh;

    @ColumnInfo(name = "department1_en")
    @SerializedName("department1_en")
    private String department1En;

    @ColumnInfo(name = "department2_zh")
    @SerializedName("department2_zh")
    private String department2Zh;

    @ColumnInfo(name = "department2_en")
    @SerializedName("department2_en")
    private String department2En;

    @ColumnInfo(name = "department3_zh")
    @SerializedName("department3_zh")
    private String department3Zh;

    @ColumnInfo(name = "department3_en")
    @SerializedName("department3_en")
    private String department3En;

    // 联系信息 (5字段)
    @ColumnInfo(name = "mobile_phone")
    @SerializedName("mobile_phone")
    private String mobilePhone;

    @ColumnInfo(name = "company_phone1")
    @SerializedName("company_phone1")
    private String companyPhone1;

    @ColumnInfo(name = "company_phone2")
    @SerializedName("company_phone2")
    private String companyPhone2;

    @ColumnInfo(name = "email")
    @SerializedName("email")
    private String email;

    @ColumnInfo(name = "line_id")
    @SerializedName("line_id")
    private String lineId;

    // 地址信息 (4字段)
    @ColumnInfo(name = "company_address1_zh")
    @SerializedName("company_address1_zh")
    private String companyAddress1Zh;

    @ColumnInfo(name = "company_address1_en")
    @SerializedName("company_address1_en")
    private String companyAddress1En;

    @ColumnInfo(name = "company_address2_zh")
    @SerializedName("company_address2_zh")
    private String companyAddress2Zh;

    @ColumnInfo(name = "company_address2_en")
    @SerializedName("company_address2_en")
    private String companyAddress2En;

    // 备注信息 (2字段)
    @ColumnInfo(name = "note1")
    @SerializedName("note1")
    private String note1;

    @ColumnInfo(name = "note2")
    @SerializedName("note2")
    private String note2;

    // 系统字段
    @ColumnInfo(name = "created_at")
    @SerializedName("created_at")
    private String createdAt;

    @ColumnInfo(name = "updated_at")
    @SerializedName("updated_at")
    private String updatedAt;

    @ColumnInfo(name = "health_status")
    @SerializedName("health_status")
    private String healthStatus; // "normal" 或 "incomplete"

    @ColumnInfo(name = "image_path")
    private String imagePath; // 本地图片路径

    // 构造函数
    public Card() {
        // 默认值为空字符串
        this.nameZh = "";
        this.nameEn = "";
        this.companyNameZh = "";
        this.companyNameEn = "";
        this.positionZh = "";
        this.positionEn = "";
        this.position1Zh = "";
        this.position1En = "";
        this.department1Zh = "";
        this.department1En = "";
        this.department2Zh = "";
        this.department2En = "";
        this.department3Zh = "";
        this.department3En = "";
        this.mobilePhone = "";
        this.companyPhone1 = "";
        this.companyPhone2 = "";
        this.email = "";
        this.lineId = "";
        this.companyAddress1Zh = "";
        this.companyAddress1En = "";
        this.companyAddress2Zh = "";
        this.companyAddress2En = "";
        this.note1 = "";
        this.note2 = "";
        this.healthStatus = "incomplete";
        this.imagePath = "";
    }

    // Getter and Setter methods
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getNameZh() {
        return nameZh != null ? nameZh : "";
    }

    public void setNameZh(String nameZh) {
        this.nameZh = nameZh != null ? nameZh : "";
    }

    public String getNameEn() {
        return nameEn != null ? nameEn : "";
    }

    public void setNameEn(String nameEn) {
        this.nameEn = nameEn != null ? nameEn : "";
    }

    public String getCompanyNameZh() {
        return companyNameZh != null ? companyNameZh : "";
    }

    public void setCompanyNameZh(String companyNameZh) {
        this.companyNameZh = companyNameZh != null ? companyNameZh : "";
    }

    public String getCompanyNameEn() {
        return companyNameEn != null ? companyNameEn : "";
    }

    public void setCompanyNameEn(String companyNameEn) {
        this.companyNameEn = companyNameEn != null ? companyNameEn : "";
    }

    public String getPositionZh() {
        return positionZh != null ? positionZh : "";
    }

    public void setPositionZh(String positionZh) {
        this.positionZh = positionZh != null ? positionZh : "";
    }

    public String getPositionEn() {
        return positionEn != null ? positionEn : "";
    }

    public void setPositionEn(String positionEn) {
        this.positionEn = positionEn != null ? positionEn : "";
    }

    public String getPosition1Zh() {
        return position1Zh != null ? position1Zh : "";
    }

    public void setPosition1Zh(String position1Zh) {
        this.position1Zh = position1Zh != null ? position1Zh : "";
    }

    public String getPosition1En() {
        return position1En != null ? position1En : "";
    }

    public void setPosition1En(String position1En) {
        this.position1En = position1En != null ? position1En : "";
    }

    public String getDepartment1Zh() {
        return department1Zh != null ? department1Zh : "";
    }

    public void setDepartment1Zh(String department1Zh) {
        this.department1Zh = department1Zh != null ? department1Zh : "";
    }

    public String getDepartment1En() {
        return department1En != null ? department1En : "";
    }

    public void setDepartment1En(String department1En) {
        this.department1En = department1En != null ? department1En : "";
    }

    public String getDepartment2Zh() {
        return department2Zh != null ? department2Zh : "";
    }

    public void setDepartment2Zh(String department2Zh) {
        this.department2Zh = department2Zh != null ? department2Zh : "";
    }

    public String getDepartment2En() {
        return department2En != null ? department2En : "";
    }

    public void setDepartment2En(String department2En) {
        this.department2En = department2En != null ? department2En : "";
    }

    public String getDepartment3Zh() {
        return department3Zh != null ? department3Zh : "";
    }

    public void setDepartment3Zh(String department3Zh) {
        this.department3Zh = department3Zh != null ? department3Zh : "";
    }

    public String getDepartment3En() {
        return department3En != null ? department3En : "";
    }

    public void setDepartment3En(String department3En) {
        this.department3En = department3En != null ? department3En : "";
    }

    public String getMobilePhone() {
        return mobilePhone != null ? mobilePhone : "";
    }

    public void setMobilePhone(String mobilePhone) {
        this.mobilePhone = mobilePhone != null ? mobilePhone : "";
    }

    public String getCompanyPhone1() {
        return companyPhone1 != null ? companyPhone1 : "";
    }

    public void setCompanyPhone1(String companyPhone1) {
        this.companyPhone1 = companyPhone1 != null ? companyPhone1 : "";
    }

    public String getCompanyPhone2() {
        return companyPhone2 != null ? companyPhone2 : "";
    }

    public void setCompanyPhone2(String companyPhone2) {
        this.companyPhone2 = companyPhone2 != null ? companyPhone2 : "";
    }

    public String getEmail() {
        return email != null ? email : "";
    }

    public void setEmail(String email) {
        this.email = email != null ? email : "";
    }

    public String getLineId() {
        return lineId != null ? lineId : "";
    }

    public void setLineId(String lineId) {
        this.lineId = lineId != null ? lineId : "";
    }

    public String getCompanyAddress1Zh() {
        return companyAddress1Zh != null ? companyAddress1Zh : "";
    }

    public void setCompanyAddress1Zh(String companyAddress1Zh) {
        this.companyAddress1Zh = companyAddress1Zh != null ? companyAddress1Zh : "";
    }

    public String getCompanyAddress1En() {
        return companyAddress1En != null ? companyAddress1En : "";
    }

    public void setCompanyAddress1En(String companyAddress1En) {
        this.companyAddress1En = companyAddress1En != null ? companyAddress1En : "";
    }

    public String getCompanyAddress2Zh() {
        return companyAddress2Zh != null ? companyAddress2Zh : "";
    }

    public void setCompanyAddress2Zh(String companyAddress2Zh) {
        this.companyAddress2Zh = companyAddress2Zh != null ? companyAddress2Zh : "";
    }

    public String getCompanyAddress2En() {
        return companyAddress2En != null ? companyAddress2En : "";
    }

    public void setCompanyAddress2En(String companyAddress2En) {
        this.companyAddress2En = companyAddress2En != null ? companyAddress2En : "";
    }

    public String getNote1() {
        return note1 != null ? note1 : "";
    }

    public void setNote1(String note1) {
        this.note1 = note1 != null ? note1 : "";
    }

    public String getNote2() {
        return note2 != null ? note2 : "";
    }

    public void setNote2(String note2) {
        this.note2 = note2 != null ? note2 : "";
    }

    public String getCreatedAt() {
        return createdAt != null ? createdAt : "";
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }

    public String getUpdatedAt() {
        return updatedAt != null ? updatedAt : "";
    }

    public void setUpdatedAt(String updatedAt) {
        this.updatedAt = updatedAt;
    }

    public String getHealthStatus() {
        return healthStatus != null ? healthStatus : "incomplete";
    }

    public void setHealthStatus(String healthStatus) {
        this.healthStatus = healthStatus != null ? healthStatus : "incomplete";
    }

    public String getImagePath() {
        return imagePath != null ? imagePath : "";
    }

    public void setImagePath(String imagePath) {
        this.imagePath = imagePath != null ? imagePath : "";
    }

    /**
     * 获取显示用的姓名（优先中文，其次英文）
     */
    public String getDisplayName() {
        if (!getNameZh().isEmpty()) {
            return getNameZh();
        }
        return getNameEn();
    }

    /**
     * 获取显示用的公司名称（优先中文，其次英文）
     */
    public String getDisplayCompanyName() {
        if (!getCompanyNameZh().isEmpty()) {
            return getCompanyNameZh();
        }
        return getCompanyNameEn();
    }

    /**
     * 获取显示用的职位（优先中文，其次英文）
     */
    public String getDisplayPosition() {
        if (!getPositionZh().isEmpty()) {
            return getPositionZh();
        }
        if (!getPositionEn().isEmpty()) {
            return getPositionEn();
        }
        if (!getPosition1Zh().isEmpty()) {
            return getPosition1Zh();
        }
        return getPosition1En();
    }

    /**
     * 获取显示用的部门（优先中文，其次英文）
     */
    public String getDisplayDepartment() {
        if (!getDepartment1Zh().isEmpty()) {
            return getDepartment1Zh();
        }
        if (!getDepartment1En().isEmpty()) {
            return getDepartment1En();
        }
        if (!getDepartment2Zh().isEmpty()) {
            return getDepartment2Zh();
        }
        return getDepartment2En();
    }

    /**
     * 获取主要联系方式
     */
    public String getPrimaryContact() {
        if (!getMobilePhone().isEmpty()) {
            return getMobilePhone();
        }
        if (!getCompanyPhone1().isEmpty()) {
            return getCompanyPhone1();
        }
        if (!getEmail().isEmpty()) {
            return getEmail();
        }
        return getLineId();
    }

    /**
     * 检查名片健康状态
     */
    public boolean isHealthy() {
        // 姓名：中英文任一有值即可
        boolean hasName = !getNameZh().isEmpty() || !getNameEn().isEmpty();

        // 公司：中英文任一有值即可
        boolean hasCompany = !getCompanyNameZh().isEmpty() || !getCompanyNameEn().isEmpty();

        // 职位或部门：至少有一个有值
        boolean hasPositionOrDept = !getPositionZh().isEmpty() || !getPositionEn().isEmpty() ||
                                   !getPosition1Zh().isEmpty() || !getPosition1En().isEmpty() ||
                                   !getDepartment1Zh().isEmpty() || !getDepartment1En().isEmpty() ||
                                   !getDepartment2Zh().isEmpty() || !getDepartment2En().isEmpty() ||
                                   !getDepartment3Zh().isEmpty() || !getDepartment3En().isEmpty();

        // 联系方式：手机/电话/Email/Line ID至少一个有值
        boolean hasContact = !getMobilePhone().isEmpty() || !getCompanyPhone1().isEmpty() ||
                            !getCompanyPhone2().isEmpty() || !getEmail().isEmpty() || !getLineId().isEmpty();

        return hasName && hasCompany && hasPositionOrDept && hasContact;
    }
}

File: backend/models/card.py
from pydantic import BaseModel
from typing import Optional, List
from sqlalchemy import Column, Integer, String, DateTime, Text, Index, ForeignKey, Float
from sqlalchemy.orm import relationship
from backend.models.db import Base
import datetime

class CardORM(Base):
    __tablename__ = "cards"
    id = Column(Integer, primary_key=True, index=True)
    
    # 基本資訊欄位（中英文）
    name_zh = Column(String(100), index=True)                 # 姓名(中文)
    name_en = Column(String(100))                             # 英文姓名
    company_name_zh = Column(String(200), index=True)        # 公司名稱(中文)
    company_name_en = Column(String(200))                     # 英文公司名稱
    position_zh = Column(String(100))                        # 職位(中文)
    position_en = Column(String(100))                        # 英文職位
    position1_zh = Column(String(255))                       # 職位1(中文)
    position1_en = Column(String(255))                       # 職位1(英文)
    
    # 部門組織架構（中英文，三層）
    department1_zh = Column(String(100))                     # 部門1(中文)
    department1_en = Column(String(100))                     # 部門1(英文)
    department2_zh = Column(String(100))                     # 部門2(中文)
    department2_en = Column(String(100))                     # 部門2(英文)
    department3_zh = Column(String(100))                     # 部門3(中文)
    department3_en = Column(String(100))                     # 部門3(英文)
    
    # 聯絡資訊
    mobile_phone = Column(String(50), index=True)            # 手機
    company_phone1 = Column(String(50))                      # 公司電話1
    company_phone2 = Column(String(50))                      # 公司電話2
    fax = Column(String(50))                                  # 傳真號碼
    email = Column(String(200))                              # Email
    line_id = Column(String(100))                            # Line ID
    wechat_id = Column(String(100))                          # WeChat ID
    
    # 地址資訊（中英文）
    company_address1_zh = Column(String(300))                # 公司地址一(中文)
    company_address1_en = Column(String(300))                # 公司地址一(英文)
    company_address2_zh = Column(String(300))                # 公司地址二(中文)
    company_address2_en = Column(String(300))                # 公司地址二(英文)
    
    # 備註資訊
    note1 = Column(Text)                          # 備註1
    note2 = Column(Text)                          # 備註2
    
    # 系統管理欄位
    front_image_path = Column(String(500))        # 正面圖片路径
    back_image_path = Column(String(500))         # 反面圖片路径
    front_ocr_text = Column(Text)                 # 正面OCR原始文字
    back_ocr_text = Column(Text)                  # 反面OCR原始文字
    created_at = Column(DateTime, default=datetime.datetime.utcnow, index=True)
    updated_at = Column(DateTime, default=datetime.datetime.utcnow, onupdate=datetime.datetime.utcnow)

    # AI 产业分类字段
    industry_category = Column(String(50), index=True)           # 产业分类：防詐/旅宿/工業應用/食品業/其他
    classification_confidence = Column(Float)                    # 分类信心度 (0-100)
    classification_reason = Column(Text)                         # 分类理由
    classified_at = Column(DateTime, index=True)                 # 分类时间

    # 複合索引，優化常見查詢
    __table_args__ = (
        Index('idx_name_company', 'name_zh', 'company_name_zh'),  # 姓名+公司複合索引
        Index('idx_name_phone', 'name_zh', 'mobile_phone'),       # 姓名+手機複合索引
    )

class Card(BaseModel):
    id: Optional[int] = None
    
    # 基本資訊欄位（中英文）
    name_zh: Optional[str] = None                 # 姓名(中文)
    name_en: Optional[str] = None                 # 英文姓名
    company_name_zh: Optional[str] = None         # 公司名稱(中文)
    company_name_en: Optional[str] = None         # 英文公司名稱
    position_zh: Optional[str] = None             # 職位(中文)
    position_en: Optional[str] = None             # 英文職位
    position1_zh: Optional[str] = None            # 職位1(中文)
    position1_en: Optional[str] = None            # 職位1(英文)
    
    # 部門組織架構（中英文，三層）
    department1_zh: Optional[str] = None          # 部門1(中文)
    department1_en: Optional[str] = None          # 部門1(英文)
    department2_zh: Optional[str] = None          # 部門2(中文)
    department2_en: Optional[str] = None          # 部門2(英文)
    department3_zh: Optional[str] = None          # 部門3(中文)
    department3_en: Optional[str] = None          # 部門3(英文)
    
    # 聯絡資訊
    mobile_phone: Optional[str] = None            # 手機
    company_phone1: Optional[str] = None          # 公司電話1
    company_phone2: Optional[str] = None          # 公司電話2
    fax: Optional[str] = None                     # 傳真號碼
    email: Optional[str] = None                   # Email
    line_id: Optional[str] = None                 # Line ID
    wechat_id: Optional[str] = None               # WeChat ID
    
    # 地址資訊（中英文）
    company_address1_zh: Optional[str] = None     # 公司地址一(中文)
    company_address1_en: Optional[str] = None     # 公司地址一(英文)
    company_address2_zh: Optional[str] = None     # 公司地址二(中文)
    company_address2_en: Optional[str] = None     # 公司地址二(英文)
    
    # 備註資訊
    note1: Optional[str] = None                   # 備註1
    note2: Optional[str] = None                   # 備註2
    
    # 系統管理欄位
    front_image_path: Optional[str] = None        # 正面圖片路径
    back_image_path: Optional[str] = None         # 反面圖片路径
    front_ocr_text: Optional[str] = None          # 正面OCR原始文字
    back_ocr_text: Optional[str] = None           # 反面OCR原始文字
    created_at: Optional[datetime.datetime] = None
    updated_at: Optional[datetime.datetime] = None

    # AI 产业分类字段
    industry_category: Optional[str] = None       # 产业分类
    classification_confidence: Optional[float] = None  # 分类信心度
    classification_reason: Optional[str] = None   # 分类理由
    classified_at: Optional[datetime.datetime] = None  # 分类时间

    model_config = {"from_attributes": True} 